### PR TITLE
feature/zagaris2/git-macros

### LINKED ISSUE #4: Add macros for basic git commands in BLT
BLT should provide a few cmake functions or macros to glean basic git information. For example, a Git the tag of a branch and/or the SHA1 at the tip of a branch.

The motivation for this is to provide the means for BLT-based projects to configure header files that consist of the Project's version, e.g., something like a "Version.h", or "configure.h"

For example, potential things client code would want to do include the following:
```
## gets the tag from the given branch, i.e., with git --describe
blt_git_tag( BRANCH <BRANCH> SOURCE_DIR <SOURCE_DIR> )
```

```
## gets the SHA1 hash at the tip of the given branch
blt_git_hash( BRANCH <BRANCH> SOURCE_DIR <SOURCE_DIR>  )
```

```
## get the name of the branch that is checked out
blt_git_branch( SOURCE_DIR <SOURCE_DIR> )
```

### PR DESCRIPTION
**Pull request by gzagaris**
Adds git macros to get basic git information such as the
tag or hashcode at the tip of a branch or the current
active branch. The motivation for this is to allow Git
based projects to dynamically encode project version
information in a generated header by leveraging Git's
native capabilities to provide this information.

A generic git macro is provided, blt_git(). This allows
running arbitrary git commands and forms the foundation
for all other macros. All macros can optionally take
a SOURCE_DIR argument to specify the path to the git
repository. If SOURCE_DIR is not specified, CMAKE_CURRENT_SOURCE_DIR
is used internally. In addition, the blt_is_git_repo() macro
is provided to faciliate checking if the source directory is
a valid Git repository or an exported tarball (i.e., no .git folder)

This commit resolves #4.

### CODE REVIEW

**Comment by cyrush** [comment]
Looks great! Can we test these in the test project?

https://github.com/LLNL/blt/tree/master/cmake/blt-test

File: cmake/thirdparty/SetupGit.cmake [hunk 1]
```diff
@@ -0,0 +1,47 @@
+###############################################################################
```
Comment (white238) [inline cmake/thirdparty/SetupGit.cmake]: Please put this directly in SetupThirdParty.cmake.  No reason to add an extra file for 5 lines.

Comment (gzagaris) [inline cmake/thirdparty/SetupGit.cmake]: Agreed. I'll make this change

File: SetupBLT.cmake [hunk 2]
```diff
@@ -87,6 +87,13 @@ if (NOT BLT_LOADED)
   include(${BLT_ROOT_DIR}/cmake/thirdparty/SetupThirdParty.cmake)
 
   ################################
+  # Git related Macros
+  ################################
+  if (Git_FOUND)
```
Comment (cyrush) [inline SetupBLT.cmake]: Can we also support, and mainly use this via all caps eg: GIT_FOUND ?

File: cmake/blt-test/CMakeLists.txt [hunk 3]
```diff
@@ -95,3 +95,61 @@ blt_add_executable(
 
 blt_add_test(NAME t_header_only_smoke
              COMMAND t_header_only_smoke)
+
+####################
+# Git Macros test
+####################
+if ( Git_FOUND )
```
Comment (cyrush) [inline cmake/blt-test/CMakeLists.txt]: checked travis, it didn't look like `t_git_macros_smoke` is being run?

Comment (gzagaris) [inline cmake/blt-test/CMakeLists.txt]: Hmm... is Git_FOUND? Here is the output from running locally:

```
zagaris2@barbspike > ctest
Test project /Users/zagaris2/dev/ASCTK/build/blt-test
    Start 1: t_example_smoke
1/4 Test #1: t_example_smoke ..................   Passed    0.02 sec
    Start 2: t_header_only_smoke
2/4 Test #2: t_header_only_smoke ..............   Passed    0.01 sec
    Start 3: t_git_macros_smoke
3/4 Test #3: t_git_macros_smoke ...............   Passed    0.01 sec
    Start 4: blt_gtest_smoke
4/4 Test #4: blt_gtest_smoke ..................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 4
```

File: SetupBLT.cmake (hunk 2 above)
Comment (gzagaris) [inline SetupBLT.cmake]: Sure

**Comment by gzagaris** [comment]
@cyrush  The reason travis is not running "t_git_macros_smoke" is because we are yanking the test project, cmake/blt-test, outside of a Git repo. See https://github.com/LLNL/blt/blob/master/.travis.yml around lines 112. Since the source is not a Git repository, we cannot run any Git commands therein. 

Perhaps, we can just do a symlink instead of copying it out, or just create a build space on Travis that points to that subfolder within the BLT repo.

We should revisit how to do this. It was awkward adding a test in `blt-test` b/c the test project expects BLT in a particular location etc. I suggest the test project requires a BLT_DIR argument when we configure with CMake. For example, something like this:

```
cmake -DBLT_DIR=</path/to/blt>  <path/to/blt>/cmake/blt-test 
```

**Comment by cyrush** [comment]
I agree the testing setup is a bit confusing - but serves the current goal of getting things tested.

An easy temp solution would be to do a git init and some commits in the test project for travis.

**Comment by gzagaris** [comment]
I actually like the "temp" solution better. That way I can hard-code what the tag is and have a more proper google-test for it in contrast to just printing that stuff out. I'll do that as part of this PR.

**Pull request merged**